### DISCUSSION TITLE
Create a class diagram for book my show

### DISCUSSION THREAD

**Discussion by kanmaytacker** [kanmaytacker/design-questions#27] [General]
What?
Identify the various entities, their attributes, behaviour and relationships with other entities.

Why?
1. Helps identifying and documenting low level implementation
2. Removes uncertainty from actual implementation

How?
Fill out the worksheet [here](https://github.com/kanmaytacker/design-questions/blob/master/notes/06-design-book-my-show-worksheet.md#class-diagrams).

**Comment by ArvindGanur** [comment]
BookMyShow
 -User
-Movie
-Theater
-Place

User
- Id
- Name
- Email
- Phno

Theater
-Name
-place
-NoOf-shows
-TypeOfSeats
-2Dor3D
-ShowTiming
-NoofScreening
Movie
-name
-Duration
-price
-reating

MovieNameDetails
-NamesOfAcotrs
-Role

TypeOfSeats
-Gold
-Diamond
-Platinum

PayMentGateWayEnum
-UPI
-creditcard
-OnlineBanking

**Reply by kanmaytacker** [reply]
A few improvements @ArvindGanur 
1. You are missing a few entities like City, Seat, Shows.
2. You have a requirement to show the available seats. How will you achieve that?
3. Also you need track user bookings as well.

**Comment by hey-vishal** [comment]
![mermaid-diagram-2023-10-18-225030](https://github.com/kanmaytacker/design-questions/assets/59616645/957089a9-b553-49e8-857c-bbcbd538a698)

i wasn't able to come up with whole it may have improvements

**Reply by kanmaytacker** [reply]
Great effort @hey-vishal.
A few things to add:
1. [MINOR] You are missing some secondary entities like payment, user, etc.
2. [MAJOR] How do you show the available and booked seats to the user?

**Reply by hey-vishal** [reply]
points noted... will do better next time... thanks @kanmaytacker

**Comment by shaileshexp** [comment]
![WhatsApp Image 2023-10-18 at 22 52 10](https://github.com/kanmaytacker/design-questions/assets/33438762/a339a0a1-0dfa-4309-b28a-bf9a204e459a)

**Reply by kanmaytacker** [reply]
Great work @shaileshexp.
You have all the entities. You might not need an extra booking entity since you have a ticket.
Something you have think about:
**How do you show available and booked seats to the user?**

**Comment by pramodpatil812** [comment]
@startuml
abstract class User {
long id
String name
}
class Customer
class Admin

class City {
int id
int name
Theatre[] theatres
}
class Theatre {
int id
String name
Hall[] halls
}
class Hall {
int id
String name
Seat[] seats
Show[] shows
}
class Seat {
int id
String number
SeatType type
}
enum SeatType {
GOLD
PLATINUM
SILVER
}
class Show {
int id
DateTime starttime
DateTime endtime
Movie movie
String language
ShowSeat[] showseats    /'redundant information as we already have this information available in ShowSeat class'/
}
class Movie {
int id
String name
Genre[] genre
String[] languages
}
enum Genre {
HORROR
THRILLER
DRAMA
}
class ShowSeat {
Show show
Seat seat
SeatStatus seatStatus
}
enum SeatStatus {
AVAILABLE
BOOKED
}
class Ticket {
int id
DateTime bookingTime
Customer customer
double price
Show show
Hall hall
Seat[] bookedSeats
Payment[] payment   /' multiple payments, if partial payments are allowed for a single ticket and if payment success happended after any failed payment '/
TicketStatus ticketStatus
}
enum TicketStatus {
BOOKED
CANCELLED
}
class Payment {
double amount
Ticket ticket
PaymentMode mode
PaymentStatus paymentStatus
}
enum PaymentMode {
UPI
CREDIT_CARD
NETBANKING
}
enum PaymentStatus {
SUCCESS
FAILED
}

/' Inheritance '/
Customer --|> User
Admin --|> User

/' Composition - City "is composed of" theatres. If City doesn't exists Theatre also can't exist '/
City *-- Theatre
Theatre *-- Hall
Hall *-- Seat
Hall *-- Show
@enduml
![bookmyshow](https://github.com/kanmaytacker/design-questions/assets/5285389/68637922-1280-4459-9d96-231284c47324)

**Comment by venu123** [comment]
![bookmyshow class diagram](https://github.com/kanmaytacker/design-questions/assets/1032681/0df89fe0-2724-47d2-aef4-c137a0638cc9)

**Comment by ershivanikhandelwal** [comment]
![image](https://github.com/kanmaytacker/design-questions/assets/139387820/95151b95-d6ee-4913-9f34-88c826bb4e0d)

**Reply by kanmaytacker** [reply]
Perfect @ershivanikhandelwal 
You have all the entities you would need. You might not want separate child classes for Cash, UPI etc.

**Something you have think about:**
How do you show available and booked seats to the user?

**Comment by animesh799** [comment]
![WhatsApp Image 2023-10-18 at 22 57 56_f852e7cf](https://github.com/kanmaytacker/design-questions/assets/93705578/2a80848c-f81d-4351-8fca-45fad40ebb28)

**Comment by vishwaraths** [comment]
![image](https://github.com/kanmaytacker/design-questions/assets/11187838/d10b0dbb-aac5-4182-9548-09fa8270e4be)

**Comment by MauryaBrij** [comment]
![image](https://github.com/kanmaytacker/design-questions/assets/62421744/07451459-fd8f-4219-a610-b2253fd44eef)

**Comment by agarwalprerna643** [comment]
![WhatsApp Image 2023-10-18 at 23 00 29_a7e69663](https://github.com/kanmaytacker/design-questions/assets/51155049/adb5616a-c63b-44fd-bd16-9b5ee1d0f12a)

**Comment by aniketankalikar** [comment]
![book_my_show](https://github.com/kanmaytacker/design-questions/assets/43496595/4f486091-0ee3-4e91-b412-f8b169f78d03)

**Comment by nikhilreddykunta** [comment]
Class City
        - String name
        
    Class Theatre
        - String name
        - City city
        
    Class Screen
        - String screenName
        - Theatre theatre
        - Movie movie
        - List<List<Seats>> seatLayout
        
    Class Movie
        - String name
        - double rating
        - List<Category> category
        
    enum Category{
    }
    
    Class Show
        - Timestamp time
        - Movie movie
    
    Class Seat
        - SeatType seatType
        - BookingStatus bookingStatus
        
    enum SeatType {
    GOLD, PLATINUM, DIAMOND
    }
    
    enum BookingStatus{
    AVAILABLE, BOOKED
    }
    
    Class Ticket
        - Screen screen
        - Movie movie
        - List<Seat> seats
    
    Class Payment
    
    Class Coupon
        - String code
    
    Class Booking
        - Ticket ticket
        - Date date
    
    
    
       
      
        ...

**Comment by annant05** [comment]
![image](https://github.com/kanmaytacker/design-questions/assets/21981463/02d4d63c-6b3e-4d80-9282-0189850f1f22)

**Reply by annant05** [reply]
classDiagram
    class City{
      id
      name
    }
    class Theatre{
      id
      name
      city_id
      isOpen()
    }
    class Screen{
      id
      movies
      isAvaliable()
    }
    class Movie{
        name
        timeslot_id
        theatre_id
    }
    class User{
        id
        name
        mobile 
        email
        created_at
    }
    class Ticket{
        seats
        user_id
        movie_id
        created_at
        payment_id
        timeslot_id
    }
    class Seat{
        SeatType
        row
        col
        isAvaliable()
    }
    class SeatType{
        name
    }

    class Payment{
        amount
        PaymentMode
    }
    class PaymentMode{
        name
    }

    Theatre <-- Ticket
    Theatre <-- Seat  
    Theatre <-- Movie
    Theatre <-- Screen
    Ticket <-- Seat
    Ticket <-- Movie
    Ticket <-- Screen
    Ticket <-- User
    Ticket <-- Payment
    Payment <-- PaymentMode
    Seat <-- SeatType
    City <-- Theatre

**Comment by rahul-ydav** [comment]
classDiagram
  class BookMyShow {
    - AppName : String
    - Owner : String
    + Cities: List<City>
    + Search(City) City
    + Book() 
  }

  class City {
    - Name: String
    - Theatres : List<Theatre>
  }

  class Theatre {
    - Id : Long
    - Name : String
    - Screens: List<Screen>
  }

  class Screen {
    - Id : Long
    - Name : String
    - Seats : List<Seat>
    - TodayShows: List<Show>
    +play(Board) BoardCell
  }

  class Seat{
    - Id: Long
    - SeatType : SeatType
    - Price : float
  }

  class Movie {
    - Name : String
    - ScreenTime : Date
    - Cast : String
    - Shows: List<Show>
  }

  class Show {
    - StartTime : Date
    - AdTimes : Date
  }

  BookMyShow "1" --* "*" City
  City "1" --* "*" Theatre
  Theatre "1" --* "*" Screen
  Screen "1" --* "*" Seat
  Movie "1" --* "*" Show
  Screen "*" --* "*" Show

**Comment by asknishant** [comment]
1.City
 - id
 - name

2.Threatre
- id
- name
- address
- screens

3.Screen
- id
- name
- seats

4.Seat
- id
- number
- row
- column

5.SeatInShow
- seatStatus
- show

SeatStatus(enum)
Booked, Empty, InProcess

6.Show
- id
- nam
- actors
- movie

7. Payment
- id
- PaymnetMode
- PaymentStatus

PaymentStatus(enum)
Cancelled, Done

8. Ticket
- id 
- show
- seat
- theatre

**Comment by perfectprocrastinator** [comment]
![20231018_230907](https://github.com/kanmaytacker/design-questions/assets/40151245/480b5539-9c41-47c5-b006-48acf99cf4cd)
